### ISSUE TITLE
ManageMembers:

### ISSUE THREAD

**Issue by klequis** [trivalleycoders-org/trivalleycoders#46]
# Manage Members form
_This issue has evolved from an 'Add Member' popup to a more full featured 'Manage Members'. Refining requirements here._

## Form Fields

**_id**
- Auto-populated by mongodb

**picture**
- Head-shot or other representation of user of user. 
- User can upload
- If user does not upload a picture they will get an identicon

**firstName**
- users first name

**lastName**
- users last name

**role**
- a short string of up to (tbd but probably 25 characters) describing the user

**index**
- integer used to sort members when they are rendered on the page

**formSort**
- integer used to sort members when they are rendered in the ManageMembers form

## Map UI to Actions
-  New Member => requestCreateMember
- Save => requestUpdateMember
- Form field change => updateMember
- Cancel => requestDeleteMember
- Close => updateShowManageMembers

## User data entry
- As user enters data the member in Redux is updated
- Once the entry is valid, validation a green check-mark will show
- If the entry is not valid, a red x will show

## Manage Members button click
1. ManageMembers form shows
1. Populates with all current members sorted by index

## Add Member button click
1. New member row appears as first row with special highlighting to set it apart from the other rows
2. Id is populated in state and in new mongodb row
4. Cursor is in the 'First Name' edit box for the new row

## User enters index number
- the new row should remain the first row with all other rows shorted by index
- this will require a hidden sort row where existing members are 0 through n matching their actual index value and the new row has hidden index of -1 such that it stays the first row.

## Reordering of index number
- If the existing members have index 0 through 3 and the user enters 0 for the new member:
  A. the user should be able to edit the other index numbers
  B. the other index numbers should automatically update (probably not)
  C. the new member row should move to its new place in the list of members. This is easy to code but perhaps not a great user experience unless done right visually.
  D. The rows can be ordered by drag and drop and the index is behind the scenes

_Ultimately, drag and drop seems the best solution using a library like [React DnD](http://react-dnd.github.io/react-dnd/docs-overview.html)_

## Save button click
1. Validation runs (see validation below)
2. Document for user is updated in database
3. Validation indicators are cleared

## Close button click
_maybe this should be 'Done'_
1. Un-saved edits are saved
2. Form closes

## Edit Button
- Each member row will have an Edit button which will put the row into edit mode
- The edit button will be the last column in the form
1. User clicks edit button
2. Check that no other member is in edit mode (or disable them, or ask to save, disable seems easiest)
3. The row is highlighted (row order does not change)
4. Once row is in edit mode the workflow is the same as Add Member

## Validation
- add validation code once functionality is fully fleshed out.

**Comment by klequis** [comment]
A little history on what I did
- new NewMemberForm is under /components now. We can move it when it is working but wanted to eliminate the variable of a different location.
- Your version of the form is /components/NewMemberform/index.jsx.old1
- I created /components/Page/NewForm to eliminate complexity so I could focus on the flow of state and api calls
- /elements is also under /components now and I think it needs to be but don't remember why. Anyway, that is as good a home for it as any.
